sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnReflector.java

```diff
@@ -95,1056 +39,288 @@
+ * Parses a {@link DoFn} and computes its {@link DoFnSignature}. See {@link #getOrParseSignature}.
  */
 public abstract class DoFnReflector {
+  private DoFnReflector() {}
 
-  private static final String FN_DELEGATE_FIELD_NAME = "delegate";
-
-  private enum Availability {
-    /** Indicates parameters only available in {@code @ProcessElement} methods. */
-    PROCESS_ELEMENT_ONLY,
-    /** Indicates parameters available in all methods. */
-    EVERYWHERE
-  }
-
-  /**
-   * Enumeration of the parameters available from the {@link ExtraContextFactory} to use as
-   * additional parameters for {@link DoFn} methods.
-   * <p>
-   * We don't rely on looking for properly annotated methods within {@link ExtraContextFactory}
-   * because erasure would make it impossible to completely fill in the type token for context
-   * parameters that depend on the input/output type.
-   */
-  private enum AdditionalParameter {
-
-    /** Any {@link BoundedWindow} parameter is populated by the window of the current element. */
-    WINDOW_OF_ELEMENT(Availability.PROCESS_ELEMENT_ONLY, BoundedWindow.class, "window") {
-      @Override
-      public <InputT, OutputT> TypeToken<?> tokenFor(TypeToken<InputT> in, TypeToken<OutputT> out) {
-        return TypeToken.of(BoundedWindow.class);
-      }
-    },
-
-    INPUT_PROVIDER(Availability.PROCESS_ELEMENT_ONLY, DoFn.InputProvider.class, "inputProvider") {
-      @Override
-      public <InputT, OutputT> TypeToken<?> tokenFor(TypeToken<InputT> in, TypeToken<OutputT> out) {
-        return new TypeToken<DoFn.InputProvider<InputT>>() {}.where(
-            new TypeParameter<InputT>() {}, in);
-      }
-
-      @Override
-      public boolean isHidden() {
-        return true;
-      }
-    },
-
-    OUTPUT_RECEIVER(
-        Availability.PROCESS_ELEMENT_ONLY, DoFn.OutputReceiver.class, "outputReceiver") {
-      @Override
-      public <InputT, OutputT> TypeToken<?> tokenFor(TypeToken<InputT> in, TypeToken<OutputT> out) {
-        return new TypeToken<DoFn.OutputReceiver<OutputT>>() {}.where(
-            new TypeParameter<OutputT>() {}, out);
-      }
+  private static final Map<Class<?>, DoFnSignature> SIGNATURE_CACHE = new LinkedHashMap<>();
 
-      @Override
-      public boolean isHidden() {
-        return true;
-      }
-    };
-
-    /**
-     * Create a type token representing the given parameter. May use the type token associated
-     * with the input and output types of the {@link DoFn}, depending on the extra
-     * context.
-     */
-    abstract <InputT, OutputT> TypeToken<?> tokenFor(
-        TypeToken<InputT> in, TypeToken<OutputT> out);
-
-    /**
-     * Indicates whether this enum is for testing only, hence should not appear in error messages,
-     * etc. Defaults to {@code false}.
-     */
-    boolean isHidden() {
-      return false;
+  /** @return the {@link DoFnSignature} for the given {@link DoFn}. */
+  static synchronized DoFnSignature getOrParseSignature(
```

Comment:
I think this was done intentionally to ensure thread safety (eg., we're adding to maps and such). It may now be the case that the invokers cache also needs some synchronization, etc.

We may be creating invokers in multiple threads on a given worker, for instance.